File: duna-core/src/main/java/io/duna/extend/PortVerticleFactory.java
/*
 * Copyright (c) 2016 Duna Open Source Project
 * Ministério do Planejamento, Desenvolvimento de Gestão
 * República Federativa do Brasil
 *
 * This file is part of the Duna Project.
 */
package io.duna.extend;

import com.google.inject.Injector;
import io.vertx.core.Verticle;
import io.vertx.core.spi.VerticleFactory;

import javax.inject.Inject;
import java.lang.reflect.Modifier;
import java.util.logging.Logger;

public class PortVerticleFactory implements VerticleFactory {

    private static final int factoryOrder = 10;

    private final Logger logger;

    private final Injector injector;

    @Inject
    public PortVerticleFactory(Logger logger, Injector injector) {
        this.logger = logger;
        this.injector = injector;
    }

    @Override
    public String prefix() {
        return "duna-port";
    }

    @Override
    public Verticle createVerticle(String verticleName, ClassLoader classLoader) throws Exception {
        Class<?> verticleClass = classLoader.loadClass(VerticleFactory.removePrefix(verticleName));

        if (!Verticle.class.isAssignableFrom(verticleClass)) {
            logger.severe(() -> verticleClass + " isn't a Verticle implementation");
            throw new IllegalArgumentException(verticleClass + " isn't a Verticle implementation.");
        }

        if (verticleClass.isInterface() || Modifier.isAbstract(verticleClass.getModifiers())) {
            logger.severe(() -> verticleClass + " should be a concrete class");
            throw new IllegalArgumentException(verticleClass + " should be a concrete class.");
        }

        logger.fine(() -> "Creating verticle for port " + verticleClass);

        return (Verticle) injector.getBinding(verticleClass).getProvider().get();
    }

    @Override
    public int order() {
        return factoryOrder;
    }

    @Override
    public boolean blockingCreate() {
        return true;
    }
}

File: duna-core/src/main/kotlin/io/duna/port/PortVerticleFactory.kt
package io.duna.port

import com.google.inject.Injector
import io.vertx.core.Verticle
import io.vertx.core.spi.VerticleFactory
import java.util.logging.Logger
import javax.inject.Inject

internal class PortVerticleFactory
  @Inject
  constructor(private val injector: Injector,
              private val logger: Logger) : VerticleFactory {

  override fun createVerticle(verticleName: String?, classLoader: ClassLoader?): Verticle {
    val verticleClass = classLoader?.loadClass(VerticleFactory.removePrefix(verticleName))

    if (!Verticle::class.java.isAssignableFrom(verticleClass)) {
      logger.severe { "${VerticleFactory.removePrefix(verticleName)} is not a valid verticle identifier." }
      throw IllegalArgumentException("$verticleName isn't a valid Verticle implementation")
    }

    logger.fine { "Creating verticle for port ${VerticleFactory.removePrefix(verticleName)}" }

    return injector.getBinding(verticleClass).provider.get() as Verticle
  }

  override fun blockingCreate(): Boolean = true

  override fun order(): Int = 1

  override fun prefix(): String = "duna-port"
}
